internal/server/ip/link.go

```diff
@@ -95,415 +84,344 @@ func (l *Link) add(linkType string, additionalArgs []string) error {
-		stdout, err := cmd.StdoutPipe()
-		if err != nil {
-			return vf, err
-		}
-
-		defer func() { _ = stdout.Close() }()
-
-		err = cmd.Start()
-		if err != nil {
-			return vf, err
-		}
-
-		defer func() { _ = cmd.Wait() }()
-
-		// Try and match: "vf 1 MAC 00:00:00:00:00:00, vlan 4095, spoof checking off"
-		reVlan, err := regexp.Compile(fmt.Sprintf(`vf %d MAC ((?:[[:xdigit:]]{2}:){5}[[:xdigit:]]{2}).*, vlan (\d+), spoof checking (\w+)`, vfID))
-		if err != nil {
-			return vf, err
-		}
-
-		// IP link command doesn't show the vlan property if its set to 0, so we need to detect that.
-		// Try and match: "vf 1 MAC 00:00:00:00:00:00, spoof checking off"
-		reNoVlan, err := regexp.Compile(fmt.Sprintf(`vf %d MAC ((?:[[:xdigit:]]{2}:){5}[[:xdigit:]]{2}).*, spoof checking (\w+)`, vfID))
-		if err != nil {
-			return vf, err
-		}
-
-		scanner := bufio.NewScanner(stdout)
-		for scanner.Scan() {
-			// First try and find VF and read its properties with VLAN activated.
-			res := reVlan.FindStringSubmatch(scanner.Text())
-			if len(res) == 4 {
-				vlan, err := strconv.Atoi(res[2])
-				if err != nil {
-					return vf, err
-				}
-
-				vf.Address = res[1]
-				vf.VLANs = append(vf.VLANs, map[string]int{"vlan": vlan})
-				vf.SpoofCheck = util.IsTrue(res[3])
-
-				return vf, err
-			}
-
-			// Next try and find VF and read its properties with VLAN missing.
-			res = reNoVlan.FindStringSubmatch(scanner.Text())
-			if len(res) == 3 {
-				vf.Address = res[1]
-				// Missing VLAN ID means 0 when resetting later.
-				vf.VLANs = append(vf.VLANs, map[string]int{"vlan": 0})
-				vf.SpoofCheck = util.IsTrue(res[2])
-
-				return vf, err
-			}
-		}
-
-		err = scanner.Err()
-		if err != nil {
-			return vf, err
-		}
-
-		return vf, vfNotFoundErr
-	}
-
-	// First try using the JSON output format as is more reliable to parse.
-	cmd := exec.Command(ipPath, "-j", "link", "show", l.Name)
-	stdout, err := cmd.StdoutPipe()
-	if err != nil {
-		return vf, err
-	}
-
-	defer func() { _ = stdout.Close() }()
-
-	err = cmd.Start()
-	if err != nil {
-		return vf, err
-	}
-
-	defer func() { _ = cmd.Wait() }()
-
-	// Temporary struct to decode ip output into.
-	var ifInfo []struct {
-		VFList []VirtFuncInfo `json:"vfinfo_list"`
-	}
-
-	// Decode JSON output.
-	dec := json.NewDecoder(stdout)
-	err = dec.Decode(&ifInfo)
-	if err != nil && err != io.EOF {
-		return vf, err
-	}
-
-	err = cmd.Wait()
+	link, err := linkByName(l.Name)
 	if err != nil {
-		// If JSON command fails, fallback to using regex match mode for older versions of ip tool.
-		// This does not support the newer VF "link/ether" output prefix.
-		return vfFindByRegex(l.Name, vfID)
-	}
-
-	if len(ifInfo) == 0 {
-		return vf, vfNotFoundErr
+		return VirtFuncInfo{}, err
 	}
 
-	// Search VFs returned for match.
-	found := false
-	for _, vfInfo := range ifInfo[0].VFList {
-		if vfInfo.VF == vfID {
-			vf = vfInfo // Found a match.
-			found = true
+	for _, vf := range link.Attrs().Vfs {
+		if vf.ID == vfID {
```

Comment:
I am not confident about this ID being the number we want.

In the iproute2 code the vf is taken from `IFLA_VF_MAC`, and that's where the netlink library puts it when setting it, but this `ID` seems to be just the index in the list of vfs returned from the kernel. 

I don't have any SR-IOV capable devices to try out if this is the same. If not then we might have to request a change on the netlink library to expose the correct value.

Reply:
You may want to request access to https://discuss.linuxcontainers.org/t/opening-up-the-linux-containers-lab-environment/22588

The lab has quite a few SR-IOV capable systems:
 - lantea (dual-port ConnectX-4, single-port Intel 10Gbit)
 - asuras (dual-port ConnectX-6)
 - entak (dual-port ConnectX-6)
 - velona (dual-port ConnectX-6)

Reply:
Of those at least lantea is known to be correctly configured at the firmware level as it does run SR-IOV daily tests for us.

Reply:
Finally managed to get around to this, it appears that the IDs and indices are in fact the same so the code should be fine as is.